### ISSUE TITLE
Support for role PIM Role Assignments

### ISSUE THREAD

**Issue by biggles007** [hashicorp/terraform-provider-azurerm#19912]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Through the Terraform AzureRM provider it isn't currently possible to assign PIM roles to Azure resources. Using AzAPI you can use the `Microsoft.Authorization/roleEligibilityScheduleRequests` resource type, but that then means all deployments required the AzAPI provider configuration.

It would be really helpful to have this natively in AzureRM to easily allow the zero trust/secure by design environments to easily be deployed.

Having separate resource type for the assignment would be beneficial, as the same API/resource type can be used to both assign a PIM role and also activate it.

### New or Affected Resource(s)/Data Source(s)

azurerm_role_eligibility_assignment

### Potential Terraform Configuration

```hcl
resource "azurerm_role_eligibility_assignment" "demo" {
  name = "string"
  scope = "string"
  principalId = "string"
  requestType = "string"
  roleDefinitionId = "string"
}
```

### References

https://learn.microsoft.com/en-us/azure/templates/microsoft.authorization/roleeligibilityschedulerequests?tabs=bicep&pivots=deployment-language-terraform

**Comment by LemurDaniel** [comment]
Hi,

is there any info about more progress regarding PIM?

It is needed by the company I currently work for, as of now we implemented a workaround by using null_resources to call the PIM-Api via terraform.

---

**Note: We already tried the AzApi-Provider which was only semi-useful.**

It worked for deploying PIM ScheduleRequests, but problems came as soon as something needed to be replaced or destroyed.
Since the 'azapi_resource' by default calls a DELETE-Request on the 'roleEligibilityScheduleRequests'-Endpoint. This isn't supported by the [roleEligiblityScheduleRequest-API](https://learn.microsoft.com/en-us/rest/api/authorization/role-eligibility-schedule-requests), it is actually expecting a PUT-Request of type 'AdminRemove'. 

![image](https://github.com/hashicorp/terraform-provider-azurerm/assets/49441790/2cd2518e-fa59-4a12-9152-e57c7575ecfd)

---

I noticed that the [azure-sdk-for-go](https://github.com/hashicorp/azure-sdk-for-go) used for the [authorization resources](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/authorization/client/client.go) in this provider doesn't yet support pim.

However the [go-azure-sdk](https://github.com/hashicorp/go-azure-sdk) used in other parts of the AzureRm-Provider already implements functions to call and create scheduleRequests, etc: 

go-azure-sdk: [Create eligiblityScheduleRequest](https://github.com/hashicorp/go-azure-sdk/blob/main/resource-manager/authorization/2020-10-01/roleeligibilityschedulerequests/method_create.go)

As already mentioned by the user above we would need resources for both schedule-Requests:
```terraform
resource "azurerm_role_eligibility_schedule_request" "demo" {
  # name                 = "string" # maybe optional as it is always a random GUID
  scope                = "string"
  principal_id         = "string"
  request_type         = "string" # AdminAssign, AdminRemove, etc.
  role_definition_name = "string"

  justification = "string"


  # Possible a block with different parameters for the different expirations

  # In my companies use the current workaround to only deploy Permanent-Assignments
  # Variant for NoExpiration.
  schedule {
    type = "NoExpiration"
  }

  # Variant for Duration
  schedule {
    type          = "NoExpiration"
    startDateTime = "YYYY-MM-DDTHH:mm:ss.0000000+02:00"
    duration      = "P90D"
  }
}

resource "azurerm_role_assignment_schedule_request" "demo" {
  # name                 = "string" # maybe optional as it is always a random GUID
  scope                = "string"
  principal_id         = "string"
  request_type         = "string" # AdminAssign, AdminRemove, etc.
  role_definition_name = "string"

  justification = "string"
}
```

As well as the PIM-Settings:
```terraform
# Possibly sperate resources for the rules, since those can only be defined on scopes and not per individual PIM-Assignment.
resource "azurerm_role_management_policy" "demo" {
  scope                = "string"
  role_definition_name = "string"

  # Maybe seperate blocks per rule
  activation_settings {
    maximum_duration = "string" # optional
    enabled_rules = [
      "MultiFactorAuthentication",
      "Justification",
      "Ticketing"
    ]

    justification_required = bool
    approval_required      = bool
  }

  assignment_settings {
    maximum_duration = "string" # Permanent or P365D, etc.
  }


  notification_settings_eligible {}
  #  .
  #  .
  #  .
  # etc.

}
```

**Comment by tim-krehan** [comment]
Hi Together :)

this topic got implemented with [v3.64.0](https://github.com/hashicorp/terraform-provider-azurerm/releases/tag/v3.64.0).
BUT:
There are still the settings missing for
- activation
- assignment
- notification

Is ther any plans to implement them as an additional resource?

**Comment by LemurDaniel** [comment]
Aside from the missing settings the Resource also doesn't seem to allow for Permanent-Assignment. 

Since with the [ScheduleRequest-API](https://learn.microsoft.com/en-us/rest/api/authorization/role-eligibility-schedule-requests/create?tabs=HTTP#type) you could also create Permanent-Assignment by providing 'NoExpiration'
